src/Controls/tests/BindingSourceGen.UnitTests/DiagnosticsTests.cs

```diff
@@ -220,6 +220,106 @@ public void Foo()
 		Assert.Equal("BSG0006", diagnostic.Id);
 		AssertExtensions.AssertNoDiagnostics(result.GeneratedCodeCompilationDiagnostics, "Generated code compilation");
 	}
+
+	[Theory]
+	[InlineData("private")]
+	[InlineData("protected")]
+	[InlineData("private protected")]
+	// https://github.com/dotnet/maui/issues/23534
+	public void ReportsWarningWhenSourceTypeIsUnaccessible(string modifier)
+	{
+		var source = $$"""
+			using Microsoft.Maui.Controls;
+
+			var foo = new Foo();
+			foo.Bar();
+
+			public class Foo
+			{
+				public void Bar()
+				{
+					var label = new Label();
+					label.SetBinding(Label.RotationProperty, (UnaccessibleClass a) => a.Value);
+				}
+
+				{{modifier}} class UnaccessibleClass
+				{
+					public int Value { get; set; }
+				}
+			}
+			""";
+
+		var result = SourceGenHelpers.Run(source);
+
+		var diagnostic = Assert.Single(result.SourceGeneratorDiagnostics);
+		Assert.Equal("BSG0007", diagnostic.Id);
+		AssertExtensions.AssertNoDiagnostics(result.GeneratedCodeCompilationDiagnostics, "Generated code compilation");
+	}
+
+	[Theory]
+	[InlineData("private")]
+	[InlineData("protected")]
+	[InlineData("private protected")]
+	// https://github.com/dotnet/maui/issues/23535
+	public void ReportsWarningWhenUnaccessibleFieldInPath(string modifier)
+	{
+		var source = $$"""
+			using Microsoft.Maui.Controls;
+
+			var foo = new Foo();
+			foo.Bar();
+
+			public class Foo
+			{
+				{{modifier}} int Value = 0;
+
+				public void Bar()
+				{
+					var label = new Label();
+					label.SetBinding(Label.RotationProperty, (Foo foo) => foo.Value);
+				}
+			}
+			""";
+
+		var result = SourceGenHelpers.Run(source);
+
+		var diagnostic = Assert.Single(result.SourceGeneratorDiagnostics);
+		Assert.Equal("BSG0008", diagnostic.Id);
+		AssertExtensions.AssertNoDiagnostics(result.GeneratedCodeCompilationDiagnostics, "Generated code compilation");
+	}
+
+	[Theory]
+	[InlineData("private")]
+	[InlineData("protected")]
+	[InlineData("private protected")]
+
+	// https://github.com/dotnet/maui/issues/23535
+	public void ReportsWarningWhenUnnaccessiblePropertyInPath(string modifier)
+	{
+		var source = $$"""
+			using Microsoft.Maui.Controls;
+
+			var foo = new Foo();
+			foo.Bar();
+
+			public class Foo
+			{
+				{{modifier}} int Value { get; set; }
+
+				public void Bar()
+				{
+					var label = new Label();
+					label.SetBinding(Label.RotationProperty, (Foo foo) => foo.Value);
+				}
```

Comment:
What would happen to a closure on a local variable, like:
```csharp
string myText = "hello";
label.SetBinding(Label.RotationProperty, (Foo foo) => myText);
```
Or maybe even weirder:
```csharp
var func = (Foo foo) => foo.Value;
label.SetBinding(Label.RotationProperty, (Foo foo) => func(foo));
```

Reply:
```
var func = (Foo foo) => foo.Value;
label.SetBinding(Label.RotationProperty, (Foo foo) => func(foo));
```

will issue `BSG0001` diagnostic as it contains function call which is not supported. 
https://github.com/dotnet/maui/blob/be0c492899fa7db0ea0cc8875690da6fa0c16dbd/src/Controls/src/BindingSourceGen/DiagnosticsFactory.cs#L19

```
string myText = "hello";
label.SetBinding(Label.RotationProperty, (Foo foo) => myText);
```

That's a good catch. We do not support this either. It will raise compile time error around the setter. In fact, we only support lambdas that can be made static at the moment. I will add a check to verify that the lambda is static and make all the tests use the static keyword.